refactored profile api request to include non-active job titles when called by pro

situatedbit opened the pull request:
This is a fix for #398 

When a professional had not created a payment account, their profile was blank because the profile API call (``/api/v1/en-US/users/{userID}?includeFullJobTitleID=-1``) did not return any profile data if the user didn't have any active job titles. This was by design.

I have altered the API call. If a professional is viewing their own profile, the call will return all job titles, even if none are active.

As a part of this, I refactored the server model ``PublicUserJobTitle``. It mostly overlapped with ``UserJobTitle``. The primary feature of ``PublicUserJobTitle`` was to return a restricted set of fields in the API call. But, it relied on an almost identical query for user job titles. I merged the query and fetch code in ``UserJobTitle``, and converted ``PublicUserJobTitle`` into an interface ``IPublicUserJobTitle``. This allows the API to fetch and return ``UserJobTitle`` objects as ``IPublicUserJobTitle``s. Only the fields in ``IPublicUserJobTitle`` are returned via the API.

In #437, I plan to modify the profile UI so professionals can see if their profiles are incomplete or not yet public. This fix merely allows professionals to see their profile.

situatedbit commented:
A simple way to test this fix is to onboard as a new professional. Before entering payment account information, set user.onboardingstep=NULL for that user to bypass the rest of the onboarding process. You should be able to go to /profile/{userID} and view the profile.

IagoSRL commented:
This post seems useful for this case: https://stackoverflow.com/questions/17123821/serialize-only-interface-properties-to-json-with-json-net

But would be a global change affecting all classes; may not be a problem though, just taking care of it.

Alternatively, we can process the result like in /users/index.cshtml to explicitely exports only the public properties (but the interface gets not useful then, except we implement the code there as a method).

situatedbit commented:
That's so strange. I thought I tested that.

Another thought I had was to make PublicUserJobTitle a class, and create conversion methods in UserJobTitle: ``ToPublicUserJobTitle`` or something similar. These would be called by the API methods to reduce the fields to those available to the public.

IagoSRL commented:
I like that solution. Basically what was thinking in the alternative I
commented, but if method is static FromUserJobTitle at PublicUserJobTitle,
you can pass in directly to list.Select to convert all list items; but
choose whatever solution you find best

El vie., 16 jun. 2017 18:49, Matt Schaefer <notifications@github.com>
escribió:

> That's so strange. I thought I tested that.
>
> Another thought I had was to make PublicUserJobTitle a class, and create
> conversion methods in UserJobTitle: ToPublicUserJobTitle or something
> similar. These would be called by the API methods to reduce the fields to
> those available to the public.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/loconomics/loconomics/pull/485#issuecomment-309072520>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBV8DFalwtek3qfS4x0yvnlZAdJtagVks5sEq29gaJpZM4N6eR7>
> .
>
-- 
Iago Lorenzo Salgueiro

situatedbit commented:
@IagoSRL I refactored public user job title from an interface to a class. I believe /users/{userID and /users/{userID}/job-profile are now working as they did before.

I had introduced a small bug in /users/{userID}/job-profile, but I've fixed that as well. It was ``GetItem`` was returning non-active user positions, but I fixed that.

IagoSRL commented:
@situatedbit It works fine right now, thanks.

Though, I did a small change, removing `FromUserJobTitles` and using `jobTitles.Select(LcRest.PublicUserJobTitle.FromUserJobTitle);` instead (make use of a built-in and less code to maintain).

situatedbit commented:
Nice! I like it.
